errors/middleware-upgrade-guide.md

```diff
@@ -0,0 +1,308 @@
+import type { NextRequest } from 'next/server'
+import { isAuthValid } from './lib/auth'
+
+export function middleware(request: NextRequest) {
+  // Example function to validate auth
+  if (isAuthValid(req)) {
+    return NextResponse.next()
+  }
+
+  return NextResponse.json({ message: 'Auth required' }, { status: 401 })
+}
+```
+
+#### After
+
+```typescript
+// middleware.ts
+import { NextResponse } from 'next/server'
+import type { NextRequest } from 'next/server'
+import { isAuthValid } from './lib/auth'
+
+export function middleware(request: NextRequest) {
+  // Example function to validate auth
+  if (isAuthValid(req)) {
+    return NextResponse.next()
+  }
+
+  const loginUrl = new URL('/login', request.url)
+  loginUrl.searchParams.set('from', request.nextUrl.pathname)
+
+  return NextResponse.redirect(loginUrl)
+}
+```
+
+## Cookies API Revamped
+
+### Summary of changes
+
+| Added                   | Removed       |
+| ----------------------- | ------------- |
+| `cookie.set`            | `cookie`      |
+| `cookie.delete`         | `clearCookie` |
+| `cookie.getWithOptions` | `cookies`     |
+
+### Explanation
+
+Based on beta feedback, we are changing the Cookies API in `NextRequest` and `NextResponse` to align more to a `get`/`set` model. The `Cookies` API extends Map, including methods like [entries](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/entries) and [values](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/entries).
+
+### How to upgrade
+
+`NextResponse` now has a `cookies` instance with:
+
+- `cookie.delete`
+- `cookie.set`
+- `cookie.getWithOptions`
+
+As well as other extended methods from `Map`.
+
+#### Before
+
+```javascript
+import { NextResponse } from 'next/server'
+import type { NextRequest } from 'next/server'
+
+export function middleware(request: NextRequest) {
+  // create an instance of the class to access the public methods. This uses `next()`,
+  // you could use `redirect()` or `rewrite()` as well
+  let response = NextResponse.next()
+  // get the cookies from the request
+  let cookieFromRequest = request.cookies['my-cookie']
+  // set the `cookie`
+  response.cookie('hello', 'world')
+  // set the `cookie` with options
+  const cookieWithOptions = response.cookie('hello', 'world', {
+    path: '/',
+    maxAge: 1000 * 60 * 60 * 24 * 7,
+    httpOnly: true,
+    sameSite: 'strict',
+    domain: 'example.com',
+  })
+  // clear the `cookie`
+  response.clearCookie('hello')
+
+  return response
+}
+```
+
+#### After
+
+```typescript
+// middleware.ts
+export function middleware() {
+  const response = new NextResponse()
+
+  // set a cookie
+  response.cookies.set('vercel', 'fast')
+
+  // set another cookie with options
+  response.cookies.set('nextjs', 'awesome', { path: '/test' })
+
+  // get all the details of a cookie
+  const [value, options] = response.cookies.getWithOptions('vercel')
```

Comment:
Tested this and it doesnt return a destructured array, rather an object
```suggestion
  const {value, options} = response.cookies.getWithOptions('vercel')
```

Reply:
cc @Kikobeats

Reply:
That's right @molebox, thanks for fixing 🙂